pkg/rpc/context.go

```diff
@@ -2719,26 +2748,224 @@ func (rpcCtx *Context) loadOrCreateConnAttempt(
-				rpcCtx.dialbackMu.m[nodeID] = nil
+	err := previousAttempt.Health()
+	if err == nil {
+		// If it completed without error then don't track the connection anymore and
+		// return success.
+		rpcCtx.dialbackMu.m[nodeID] = nil
+		return nil
+	}
+	// If still don't know the outcome of the previous attempt, allow this attempt
+	// to continue and check again in the future.
+	if errors.Is(err, ErrNotHeartbeated) {
+		return nil
+	}
+
+	return err
+}
+
+// hasSiblingConn takes a peer as identified by `self` and tries to find a sibling:
+// - if self.NodeID != 0, finds another conn with matching NodeID but different TargetAddr.
+// - if self.NodeID == 0, finds another conn with matching TargetAddr.
+//
+// In both cases, if such a conn exists that became healthy *after* ours became
+// unhealthy, `healthy` will be true. If no such conn exists, (false, false) is
+// returned.
+//
+// TODO(during review): move after dust has settled.
+func hasSiblingConn(peers map[peerKey]*peer, self peerKey) (healthy, ok bool) {
+	for other, otherPeer := range peers {
+		if self == other {
+			continue // exclude self
+		}
+		otherSnap := otherPeer.snap() // this would deadlock if called on `self`
+
+		if self.NodeID == 0 {
+			if other.TargetAddr != self.TargetAddr {
+				continue
 			}
-			return err
-		default:
-			// We still don't know the outcome of the previous attempt. For now
-			// allow this attempt to continue and check in the future.
-			return nil
+			// We're a GRPCUnvalidatedDial, which has a tricky life cycle because no
+			// NodeID is associated. We can't detect an IP address change in the same
+			// way we can for the NodeID!=0 branch below, nor do we get an event
+			// telling is we're decommissioned.
+			//
+			// We do the simple thing: if there's any another connection to the same
+			// address (no matter who) and it's healthy, report that. The caller can
+			// figure out what to do with that information.
+		} else {
+			if self.NodeID != other.NodeID || self.TargetAddr == other.TargetAddr {
+				continue
+			}
+			// We're a validated (i.e. with NodeID) connection and found another
+			// healthy peer matching our NodeID but not our address. Node <NodeID> has
+			// restarted under a new IP! We are likely obsolete.
+			//
+			// Note that if "we" are actually the newer connection (hypothetically
+			// this could happen if the other connection is "still" healthy, and we
+			// are "temporarily" unhealthy) then we'll remove the wrong connection,
+			// but there will be a reconnection attempt (recreating the myKey peer),
+			// so while causing a hiccup it wouldn't wedge anything.
+		}
+		ok = true
+		healthy = healthy || otherSnap.c.Health() == nil
+	}
+	return healthy, ok
+}
+
+// TODO(during review): move after dust has settled.
+func shouldDeleteAfter(conns *peerMap, myKey peerKey, err error) time.Duration {
+	conns.mu.RLock()
+	defer conns.mu.RUnlock()
+
+	sibHealthy, ok := hasSiblingConn(conns.mu.m, myKey)
```

Comment:
Is it ok that we could wait for sibling connection to get its health resolved here? Wouldn't that delay resolving gatedCC below.

Reply:
Maybe `rpc.Connection.wait` needs a better name then because that got me confused since I didn't look on the implementation. Something like `waitOrDefault`?

Reply:
Good idea, done.